oauthlib/oauth2/rfc6749/grant_types/base.py

```diff
@@ -17,6 +18,75 @@ class GrantTypeBase(object):
     error_uri = None
     request_validator = None
     default_response_mode = 'fragment'
+    refresh_token = True
+    response_types = ['code']
+
+    def __init__(self, request_validator=None, **kwargs):
+        self.request_validator = request_validator or RequestValidator()
+
+        # Transforms class variables into instance variables:
+        self.response_types = self.response_types
+        self.refresh_token = self.refresh_token
+
+        self._setup_validator_hooks()
+        for kw, val in kwargs.items():
+            setattr(self, kw, val)
+
+    def _setup_validator_hooks(self):
+        self._auth_validators_run_before_standard_ones = []
+        self._auth_validators_run_after_standard_ones = []
+        self._token_validators_run_before_standard_ones = []
+        self._token_validators_run_after_standard_ones = []
+        self._code_modifiers = []
+        self._token_modifiers = []
+
+    def register_response_type(self, response_type):
+        self.response_types.append(response_type)
+
+    def register_authorization_validator(self, validator, after_standard=True):
+        """
+        Register a validator callable to be invoked during calls to the
+        authorization endpoint.
+
+        :param callable validator: callable that takes a request object and returns a
+                          mapping of items to be included with the authorization validation
+                          response.
+        :param bool after_standard: default: True, If True, the custom
+                                    validator is called after the standard authorization
+                                    validations. If False, the custom valdator is called before
+                                    the standard validations.
+        :returns: None
+        """
+        if after_standard:
+            self._auth_validators_run_after_standard_ones.append(validator)
+        else:
+            self._auth_validators_run_before_standard_ones.append(validator)
+
+    def register_token_validator(self, validator, after_standard=True):
```

Comment:
I really went back and forth on this. I agree the boolean is a little bit clunky.

Other options:

**Use two separate methods (four when you count auth validators and token validators).**

The reasons I didn't do this were:

1) Using the boolean let me preserve backwards compatibility for anyone who is already using `register_*_validator()` in its existing incarnation.

2) Many users just want some custom validations in the mix, and don't care about whether they run before or after the existing ones. Having a boolean with a default spares those users from having to think about it.

3) I couldn't think of method names that were informative without being so verbose it seemed unwieldy, e.g. `register_auth_validator_run_before_standard()` (before standard what?)

**Just completely get rid of these registration methods and let callers append or insert to the list of validators directly.**

I actually kind of like this option, because it feels more Pythonic doing away with setter methods, but I didn't do it because it seemed out of step with the rest of the code base, and obviously it somewhat locks you into an implementation for the future.